usersession/agent/rest_api.go

```diff
@@ -197,3 +209,112 @@ func postServiceControl(c *Command, r *http.Request) Response {
+	if err != nil {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 500,
+			Result: &errorResult{
+				Message: fmt.Sprintf("cannot connect to the session bus: %v", err),
+			},
+		})
+	}
+	defer conn.Close()
+
+	// TODO: support desktop-specific notification APIs if they provide a better
+	// experience. For example, the GNOME notification API.
+	notifySrv := notification.New(conn)
+
+	// TODO: this message needs to be crafted better as it's the only thing guaranteed to be delivered.
+	summary := fmt.Sprintf(i18n.G("Pending update of %q snap"), refreshInfo.InstanceName)
+	var urgencyLevel notification.Urgency
+	var body string
+	var icon string
+	var hints []notification.Hint
+	if daysLeft := int(refreshInfo.TimeRemaining.Truncate(time.Hour).Hours() / 24); daysLeft > 0 {
+		urgencyLevel = notification.LowUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d day left)",
+			"Close the app to avoid disruptions (%d days left)", daysLeft), daysLeft)
+	} else if hoursLeft := int(refreshInfo.TimeRemaining.Truncate(time.Minute).Minutes() / 60); hoursLeft > 0 {
+		urgencyLevel = notification.NormalUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d hour left)",
+			"Close the app to avoid disruptions (%d hours left)", hoursLeft), hoursLeft)
+	} else if minutesLeft := int(refreshInfo.TimeRemaining.Truncate(time.Minute).Minutes()); minutesLeft > 0 {
+		urgencyLevel = notification.CriticalUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d minute left)",
+			"Close the app to avoid disruptions (%d minutes left)", minutesLeft), minutesLeft)
+	} else {
+		summary = fmt.Sprintf("Snap %q is refreshing now!", refreshInfo.InstanceName)
+		urgencyLevel = notification.CriticalUrgency
+	}
+	hints = append(hints, notification.WithUrgency(urgencyLevel))
+	if refreshInfo.BusyAppDesktopEntry != "" {
+		hints = append(hints, notification.WithDesktopEntry(refreshInfo.BusyAppDesktopEntry))
+		// Extract the icon manually
+		parser := goconfigparser.New()
+		if err := parser.ReadFile(fmt.Sprintf("/var/lib/snapd/desktop/applications/%s.desktop", refreshInfo.BusyAppDesktopEntry)); err == nil {
+			icon, _ = parser.Get("Desktop Entry", "Icon")
+			fmt.Printf("found desktop file icon %q\n", icon)
+		}
+	}
+	msg := &notification.Message{
+		AppName:       refreshInfo.BusyAppName,
+		Summary:       summary,
+		Icon:          icon,
+		Body:          body,
+		ExpireTimeout: time.Second * 10,
+		Hints:         hints,
+	}
+
+	// TODO: if snap store is installed and actions are supported, add an action
+	// to open the snap store page for the given snap.
+	//
+	// XXX: how are instances supported in the snap store, are they?
+
+	// TODO: silently ignore error returned when the notification server does not exist.
+	// TODO: track returned notification ID and respond to actions, if supported.
+	if _, err := notifySrv.SendNotification(msg); err != nil {
```

Comment:
We could also track the returned notification ID for as long as userd is alive, and use it to update existing notification, if one exists. This would ensure that the roster of persistent notifications only shows the current, up-to-date entry for each snap, not a separate entry per notification.

Reply:
This is one of the nicer parts of the the new GNOME notification spec: it has client provided notification IDs, so we don't need to track server IDs to update/replace/withdraw old notifications.